### ISSUE TITLE
Create a way for test runners to provide the project root of the test target

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/flutter#143170]
This issue applies to running both Dart and Flutter tests (`dart test`, and `flutter test`), but I'm filing in the Flutter repo to at least track the discussion somewhere.

**Background:**
DevTools extensions may need to connect to a test target to provide tooling (e.g. [package:patrol](https://pub.dev/packages/patrol) which already provides a DevTools extension, and an in-progress extension for [package:rohd](https://pub.dev/packages/rohd). This is possible right now if a test target is run via `dart run` or `flutter run`, but is broken when a test is run via `dart test` or `flutter test` (tracked at https://github.com/flutter/devtools/issues/7183).

In order to detect which DevTools extensions are available for a VM service connection, we need to know what the package root is for the debug target (the directory that contains the `.dart_tool/package_config.json` file). We try to determine this by looking up the root library for the main isolate: https://github.com/flutter/devtools/blob/master/packages/devtools_app/lib/src/extensions/extension_service.dart#L198.

- For `dart run` and `flutter run` this properly resolves to the target library (e.g. `file:///path/to/my_test.dart`).
- For `dart test` and `flutter test`, this resolves to a file path like: `file:///var/folders/gf/14ctmb656m3100h5z857669800fv4z/T/flutter_tools.9i1YVt/flutter_test_listener.ASUv35/listener.dart`. In this case, we can't find the project root of `my_test.dart`.

See this [thread](https://discord.com/channels/608014603317936148/1204148234960900206) on the #devtools-extension-authors Discord channel for the full discussion.

**Proposed solution:**
@jakemac53  proposed having the test runners (both for `dart test` and `flutter test`) register a service extension that returns the package root of the test target. This service extension should be available to call from DevTools, meaning it needs to be registered on the isolate we have access to (i.e. the isolate with the root library `file:///var/folders/gf/14ctmb656m3100h5z857669800fv4z/T/flutter_tools.9i1YVt/flutter_test_listener.ASUv35/listener.dart` in the example above). 

We'd also need a way from DevTools to determine when we should call this service extension. In other words, we need to be able to detect when the main isolate was spawned by `dart test` or `flutter test`. We may be able to determine this solely by looking to see whether this new service extension is registered, and if it is, we can assume we are debugging a test target.

Tagging team members from the Discord discussion: @jakemac53 @christopherfujino @derekxu16 @bkonyi @DanTup

**Comment by kenzieschmoll** [comment]
The draft at https://github.com/dart-lang/test/pull/2184/ addresses this issue `dart test`, but not for `flutter test`. 

When running `dart test test/foo_test.dart --pause-after-load`, a VM service connection is available to debug the test runner where we have registered the service extension.

When running `flutter test test/foo_test.dart --start-paused`, the VM service connection that is available does not have access to the isolate where the service extension was registered. There appears to be another layer in flutter_tools (`flutter test`) that is not present in the `dart test` case. @christopherfujino

**Comment by kenzieschmoll** [comment]
Update: with a combination of https://github.com/dart-lang/test/pull/2184 and https://github.com/flutter/flutter/pull/143481, this is the current state.

- `dart test test/foo_test.dart --pause-after-load`: DevTools can successfully detect the service extension registered by package:test (https://github.com/dart-lang/test/pull/2184) and call it. This is because the VM service connection contains the isolate where this service extension was registered.
- `flutter test test/foo_test.dart --start-paused`
  - Initially, DevTools cannot detect the service extension because the service extension will not get registered until you unpause.
  - After setting a breakpoint in the test library (a hack to ensure the process does not exit when hitting resume - we wouldn't need to do this if we had a [--pause-on-exit](https://github.com/flutter/flutter/issues/125993) flag), hit resume, and now the service extension is detectable and can be called. This is problematic because the service extension should be available when DevTools first connects.

A couple possible solutions:
1) register the service extension in some other place that is called on the test VM before it gets paused (is there a better place to perform the registration? @jonahwilliams)
2) Modify `--start-paused` flag behavior to still run some bootstrap code for the test before pausing so that the service extension will be registered
3) change DDS so that you can have it run a callback on start to optionally register service extensions that need to be available at startup. Here is a related request for being able to set initial service extension values on start up (https://github.com/dart-lang/sdk/issues/44611).

I think option one or two would be simpler. Thoughts @jonahwilliams @bkonyi @christopherfujino?

**Comment by jonahwilliams** [comment]
Maybe the vm service is not the way to do this? or alternatively, is there a way that flutter tools could provide some metadata to the vm service or dds when it starts it up?

**Comment by bkonyi** [comment]
Unfortunately, I don't think any solution with a service extension is going to work if the isolate it's registered on is paused since service extensions will only execute on running isolates.

It's hardly ideal, but the best way to find the project root for a test target will probably involve making some assumptions about how the actual test is run from the bootstrapping code in the root library. Since the root library will be importing the actual library we're interested in, we can look at the list of dependencies through [`Library.dependencies`](https://pub.dev/documentation/vm_service/latest/vm_service/Library/dependencies.html) to find it. The code probably always has the following structure:

```dart
// @dart=3.4
import "dart:isolate";

import "package:test_core/src/bootstrap/vm.dart";

import "file:///private/tmp/foo/test/foo_test.dart" as test; // This is the library we're interested in

void main(_, SendPort sendPort) {
  internalBootstrapVmTest(() => test.main, sendPort);
}
```

Looking for the dependency where [`LibraryDependency.prefix`](https://pub.dev/documentation/vm_service/latest/vm_service/LibraryDependency/prefix.html) is equal to 'test' should be good enough to figure out the library actually running the test. Assuming this prefix is always going to be 'test', this is probably the best approach, but someone more familiar with `package:test` is going to have to confirm whether or not that's the case.

**Comment by jakemac53** [comment]
Tests can be in any directory

**Comment by bkonyi** [comment]
> Tests can be in any directory

I'm not sure that matters in this case. If the import prefix is always 'test', we should be able to find the library.

**Comment by jakemac53** [comment]
Oh sorry, I misunderstood what you were saying. Generally, we wouldn't consider it breaking to change that prefix, it is not a part of the public API. So I would definitely prefer that we not rely on it. The fact that we even bootstrap tests at all like this this, or any specifics of it, are private.

**Comment by kenzieschmoll** [comment]
I'm picking this work back up. Sounds like there are drawbacks / concerns with trying to use service extensions and to trying to detect based on the test bootstrapping code, like @bkonyi suggested.

@jakemac53 is there a way we could use an environment variable that could be set from the test runner code (ideally in a way that would work for both `dart test` and `flutter test`) and then read that environment variable from DevTools?

**Comment by jakemac53** [comment]
I don't know a whole lot about environment variables, but I think a process can only set the environment for its subprocesses? At least via Dart APIs? So, I don't know that this is feasible since test doesn't actually spawn DevTools itself.

Please correct me if I am wrong/misunderstanding though!

**Comment by jakemac53** [comment]
Note that we also can't set `-D` Dart environment variables (I just realized I am not sure if you meant system or Dart environment variables, I was talking about system ones).

But in any case, in some modes we spawn tests as an isolated which cannot afaik have custom Dart environment variables.

**Comment by jakemac53** [comment]
> Open Question: can we modify dart test to set the --mark-main-isolate-as-system-isolate flag in the VM? Users will not pass this flag to run their tests, and I think it is unreasonable to expect them to in order to use developer tools.

Will that also happen for `dart run test` and `dart pub run test`? I don't know how much these commands are just aliases for each other or what.

**Comment by kenzieschmoll** [comment]
> I think we can, it'll just involve some annoying plumbing in the short term...  This won't be a problem once we migrate DartDev to AOT since we'll be spawning child processes for `dart run` and `dart test` and can easily add the `--mark-main-isolate-as-system-isolate` flag to the argument list.

If this is more work than is justified pre-moving to DartDev to AOT, another option for the short term is to handle this in two separate ways for `flutter test` and `dart test`. 
1. `flutter test`: for this case, we can grab the root library of the main isolate and look for the library import containing the test target.
2. `dart test`:
- short term: use the service extension solution https://github.com/dart-lang/test/pull/2184 and call this service extension from DevTools when we cannot find the import containing the test in the main isolate's root library available.
- long term: add the `--mark-main-isolate-as-system-isolate` flag for `dart test` once DartDev is moved to AOT

We'd still want test coverage for the `flutter test` case where we are poking at a library and making assumptions about its imports, which would eventually become the long term solution for `dart test` as well.

Thoughts?

**Comment by kenzieschmoll** [comment]
> How does this work with --pause-after-load? The service extension will be on an isolate that's paused, won't it?

Somehow the service extension gets registered from `package:test` and we can call it from DevTools... maybe it is registered on an isolate that is not paused? @jakemac53  may know the specifics.

> I think it's probably just worth doing the work to pass the flag back to the VM since there's no ETA for the DartDev migration to AOT.

If you think this is the right approach, then this is fine with me. This will simplify things on our side anyway to have a single way to grab the test target that works for both `flutter test` and `dart test`. Do you have any cycles to help with this?

**Comment by kenzieschmoll** [comment]
This issue is resolved. With a combination of Ben's [change](https://github.com/dart-lang/sdk/commit/5f3cd4afe3429338bc0670046e2b08d54e47b1ab) and this [DevTools PR](https://github.com/flutter/devtools/pull/7717), we are able to find the URI of the test target, from which we can detect the project root using existing logic. This solution will break if the import or the prefix containing the test library from Flutter tools or package:test changes. To prevent future regressions, we will:
1. Add a comment to the code that generates the 'test' prefix for both the `flutter test` and `dart test` code paths.
2. Add an integration test in DevTools that verifies we can detect extensions for Dart and Flutter test targets (https://github.com/flutter/devtools/issues/7748).
3. Add a test to package:test and flutter_tools to verify the prefix does not change.

**Comment by kenzieschmoll** [comment]
@jakemac53 do you have a code pointer for where the best place is to address items #1 and #3 from my previous comment?

**Comment by jakemac53** [comment]
This looks like the test bootstrapping code you are trying to find.

This is done in many different places and tools, specifically for the platform it is being compiled to, etc. Depending on specific contents is probably going to be highly problematic, because we allow people to "bring their own" bootstrapping too (build_test does this, and google3 does its own).

**Comment by jakemac53** [comment]
The specific URI you will see here is also likely going to vary by platform and not always be something useful.

**Comment by kenzieschmoll** [comment]
> The specific URI you will see here is also likely going to vary by platform and not always be something useful.

Can you provide an example of a case where the URI would not be useful? Do define what "useful" means for our case, we are looking for a URI from which we can detect the package root of the test target library. We detect this by walking up the tree until we find the folder that contains the `.dart_tool` directory, or by using a [regexp check](https://github.com/flutter/devtools/blob/master/packages/devtools_shared/lib/src/utils/file_utils.dart/#L81-L82) if the walk fails.

**Comment by jakemac53** [comment]
> Can you provide an example of a case where the URI would not be useful?

Specifically the `org-dartlang-app://` or other multi-root URI cases. These are special URIs which have their own "root" (the thing you actually want, but the point of them is to remove it from the URI so we have deterministic output).

So for a test you would see something like `org-dartlang-app://test/some_test.dart` which wouldn't afaik help you, because you don't know what `org-dartlang-app` means (there was a special flag to the compiler, to tell it what that means).

**Comment by jakemac53** [comment]
Note that it is quite possible in the test runner itself we don't today use these particular URIs, but we might want to in the future, and this would be a frustrating thing to be blocked on.

Google3 almost certainly does use these URIs, and build_test uses a relative URI (which I think won't help you).

**Comment by kenzieschmoll** [comment]
> Specifically the org-dartlang-app:// or other multi-root URI cases. These are special URIs which have their own "root" (the thing you actually want, but the point of them is to remove it from the URI so we have deterministic output).
> 
> So for a test you would see something like org-dartlang-app://test/some_test.dart which wouldn't afaik help you, because you don't know what org-dartlang-app means (there was a special flag to the compiler, to tell it what that means).

For this case, how can we give tooling access to the true value of `org-dartlang-app://`?

**Comment by jakemac53** [comment]
> For this case, how can we give tooling access to the true value of `org-dartlang-app://`?

This is where a service extension or some other communication that is more explicit from the test runner about the project root would work better.

> Also, general question about this logic: why do the URIs that use `org-dartlang-app://` need to be deterministic?

Specifically for hermetic build systems like build_runner/bazel this matters (the result should be the same regardless of which computer the build runs on).

But also, for release builds it is better not to have the file system structure of the system the binary was built on be a part of the compiled app. Tests are different from release builds, but it is still generally not useful information (on a CI test failure you don't really want the full stack trace just the parts relative to the package root).

**Comment by natebosch** [comment]
> we are looking for a URI from which we can detect the package root of the test target library

We do not currently have a restriction that a test suite library is in a directory under the package root. This will be true by convention in the vast majority of cases, but it is not a structural guarantee.

If we want package information, we should be using the package config. Can you read the package config information from the isolate?

**Comment by natebosch** [comment]
I suppose the package config doesn't track which package is the root  - I can't recall how reliably the `package_config.json` file is under the root package - @jakemac53 do you know if that's always the case?

We might want to revisit the idea of storing the working directory at startup.

**Comment by jakemac53** [comment]
> I suppose the package config doesn't track which package is the root - I can't recall how reliably the `package_config.json` file is under the root package - @jakemac53 do you know if that's always the case?

The package config could be anywhere, and definitely in some cases might be just under some random temp or build directory.

**Comment by natebosch** [comment]
> we need to know what the package root is for the debug target (the directory that contains the `.dart_tool/package_config.json` file)

I guess I didn't stop to think if we were looking for that directory just for that file...

@kenzieschmoll - Is finding the root directory an indirect way to find `package_config.json`? Or is there another file that you will need to look for in `.dart_tool`?

If we only need to find `package_config.json`, I think the VM should already have that information. If we are looking to read some other file, it probably won't be in `.dart_tool` whenever `package_config.json` isn't in that location.

**Comment by kenzieschmoll** [comment]
> If we want package information, we should be using the package config. Can you read the package config information from the isolate?

How can we do this? I see this information is [available from dart:isolate](https://api.flutter.dev/flutter/dart-isolate/Isolate/packageConfig.html), but not from [package:vm_service](https://github.com/dart-lang/sdk/blob/main/pkg/vm_service/lib/src/vm_service.dart#L5558). @bkonyi can we expose this?

For detecting runtime extensions (which is what we are trying to do here for `dart test` and `flutter test` run targets), we are trying to find the package root for the sole purpose of finding the `package_config.json` file: https://github.com/flutter/devtools/blob/master/packages/devtools_shared/lib/src/extensions/extension_manager.dart/#L147-L151. So if we could get this from the main isolate, we wouldn't need to do any special detection for `dart test` and `flutter test` that is different from `dart run` and `flutter run`.

**Comment by bkonyi** [comment]
> > If we want package information, we should be using the package config. Can you read the package config information from the isolate?
> 
> How can we do this? I see this information is [available from dart:isolate](https://api.flutter.dev/flutter/dart-isolate/Isolate/packageConfig.html), but not from [package:vm_service](https://github.com/dart-lang/sdk/blob/main/pkg/vm_service/lib/src/vm_service.dart#L5558). @bkonyi can we expose this?

I'm not sure this is always populated, particularly for Flutter applications. I tried access this property with a Flutter desktop application and got the following exception:

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Unsupported operation: Isolate.packageConfig
#0      Isolate.packageConfigSync (dart:isolate-patch/isolate_patch.dart:326:7)
#1      Isolate.packageConfig (dart:isolate-patch/isolate_patch.dart:319:25)
#2      main (package:gallery/main.dart:28:23)
#3      _runMain.<anonymous closure> (dart:ui/hooks.dart:301:23)
#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

There's a similar exception thrown for Flutter web.

AFAIK, the VM only cares about the package config to pass it to the CFE, which needs it for package resolution. In the case of Flutter, the VM doesn't host the CFE itself so it's not necessary for the package config to be populated. Maybe there's an argument to be made for including the package config URI in the kernel format so it's always accessible?

**Comment by jakemac53** [comment]
Yeah, in general there are a lot of contexts where a package config might not be available to a running application - and I don't think that it makes sense for it to be generally available. Once an app is compiled, the package config is no longer relevant, may not longer exist at all, or the URI might not be valid at all in the context in which the app is running (web, a different machine, etc).

**Comment by natebosch** [comment]
If we think this information doesn't make sense to expose from the VM service generally, but does make sense to expose from the test runner, I think it would be OK to register a VM service extension that proxies the `dart:isolate` APIs in the test runner. I do think that for the places we use the test runner, the VM's understanding of the package config is going to be as reliable or more than anything we can build in the test runner itself.

**Comment by kenzieschmoll** [comment]
This issue has been resolved: https://github.com/dart-lang/test/issues/2246.
